### PR TITLE
Stripe Integration for Brave Together Subscriptions / Order Fulfillment

### PR DESCRIPTION
**Pull request by dlipeles**
Implements Stripe Integration for Brave Together Subscriptions / Order Fulfillment 

#### TBD:

- [ ] Customize Stripe checkout session
- [ ] Set redirect pages from Stripe checkout session
- [ ] Add merchant to payload in credential redemption
- [ ] Set env. vars for `STRIPE_SECRET` and `STRIPE_WEBHOOK_SECRET`

#### TODO:

- [x] Set feature flags where needed
- [x] Cancellation flow for subscriptions
- [x] Add code to match SKU to Stripe price object 
- [x] Add rate limiter exception for development
- [x] Set CORS policy for brave together development
- [x] Check for existing customer before creating 
- [x] Store checkoutSessionId and SubscriptionId on Order at appropriate time
- [x] Add `payment_metadata` jsonb column to Order
- [x] Decide price for Paid Service and create new SKU token
- [x] Update `IsStripePayable()` with new SKU token

#### e2e Testing with Amir & Marshall 

- [ ] Order Creation
- [ ] Order Retrieval
- [ ] Order Payment + Subscription Handling
- [ ] Credential Signing
- [ ] Credential Verification
- [ ] Credential Reset
- [ ] Credential / Order Cancellation

#### 2. Retrieve the order; if the order is pending and Stripe payable, server will append a checkout session ID
Screen Shot 2020-11-15 at 7 26 35 PM

#### 3. Proceed to the Stripe checkout session, and enter payment credentials
Screen Shot 2020-11-15 at 7 22 45 PM

#### 4. On success, webhook is accepted by server and order is fulfilled (Set to paid)
Screen Shot 2020-11-15 at 7 27 36 PM

### CODE REVIEW

File: payment/order.go [hunk 1]
```diff
@@ -140,3 +143,61 @@ func CreateOrderItemFromMacaroon(sku string, quantity int) (*OrderItem, error) {
...
+// Create a Stripe Checkout Session for an Order
+func (order Order) CreateCheckoutSession() CreateCheckoutSessionResponse {
+	stripe.Key = os.Getenv("STRIPE_SECRET")
+
+	// Prepare checkout line items
+	lineItems := make([]*stripe.CheckoutSessionLineItemParams, len(order.Items))
+	for i := 0; i < len(order.Items); i++ {
+		price := ConvertPriceToUnitAmount(order.Items[i].Price)
+		quantity := int64(order.Items[i].Quantity)
+		name := order.Items[i].SKU
```
Comment (dlipeles) [inline payment/order.go:165]: Note to self, use friendlier descriptor than SKU

File: payment/controllers.go [hunk 2]
```diff
@@ -242,10 +251,27 @@ func GetOrder(service *Service) handlers.AppHandler {
...
 		}
 
+		// If order is pending and is stripe payable, return a checkout session
+		if !order.IsPaid() && order.IsStripePayable() {
+			type OrderWithCheckoutSession struct {
+				*Order
+				CheckoutSession CreateCheckoutSessionResponse
+			}
+
+			checkoutSession := order.CreateCheckoutSession()
```
Comment (dlipeles) [inline payment/controllers.go:261]: Note to self: need better error handling here

**Comment by dlipeles** [comment]
@evq @amir - early draft of stripe integration, I'm sure there is more we will want to work out in the e2e integration with subscriptions service, but i'm hopeful this could at least serve as a starting point

File: payment/order.go [hunk 3]
```diff
@@ -140,3 +143,61 @@ func CreateOrderItemFromMacaroon(sku string, quantity int) (*OrderItem, error) {
...
+	}
+
+	params := &stripe.CheckoutSessionParams{
+		PaymentMethodTypes: stripe.StringSlice([]string{
+			"card",
+		}),
+		Mode:              stripe.String(string(stripe.CheckoutSessionModePayment)),
+		LineItems:         lineItems,
+		ClientReferenceID: stripe.String(order.ID.String()),
+		SuccessURL:        stripe.String("https://example.com/success"),
```
Comment (dlipeles) [inline payment/order.go:185]: Note to self: we will need to configure these with urls from the client side

File: payment/order.go [hunk 4]
```diff
@@ -140,3 +143,61 @@ func CreateOrderItemFromMacaroon(sku string, quantity int) (*OrderItem, error) {
 func (order Order) IsPaid() bool {
 	return order.Status == "paid"
 }
+
+// IsStripePayable returns true if every sku is payable by Stripe
+func (order Order) IsStripePayable() bool {
+	return true
```
Comment (dlipeles) [inline payment/order.go:149]: This function will check for the new Brave Together paid SKU

File: payment/controllers.go [hunk 5]
```diff
@@ -50,6 +52,13 @@ func CredentialRouter(service *Service) chi.Router {
 	return r
 }
 
+// PaymentRouter handles calls relating to payments
+func PaymentRouter(service *Service) chi.Router {
+	r := chi.NewRouter()
+	r.Method("POST", "/stripe/webhooks", middleware.InstrumentHandler("HandleStripeWebhook", HandleStripeWebhook(service)))
```
Comment (maikelmclauflin) [inline payment/controllers.go:58]: does it make sense to have webhooks be the base, just so we don't end up sprinkling implementations around as we add more?

File: payment/controllers.go [hunk 6]
```diff
@@ -595,6 +621,47 @@ func MerchantTransactions(service *Service) handlers.AppHandler {
...
+				return handlers.WrapError(err, "error retrieving client reference ID", http.StatusBadRequest)
+			}
+			err = service.Datastore.UpdateOrder(orderID, "paid")
+			if err != nil {
+				return handlers.WrapError(err, "error updating order status", http.StatusInternalServerError)
+			}
+			return handlers.RenderContent(r.Context(), "checkout successful", w, http.StatusOK)
+		}
+
+		return handlers.RenderContent(r.Context(), "error in checkout session", w, http.StatusBadRequest)
```
Comment (husobee) [inline payment/controllers.go:661]: @dlipeles : stripe will repeat the webhook call in non OK response case... if the event.Type is not "checkout.session.completed" do we want stripe to attempt the webhook again and again for days?  I think I would suggest for the event.Type's that we dont support we just pass back an OK status.

File: payment/order.go (hunk 4 above)
Comment (husobee) [reply payment/order.go:149]: Are we going to have a list of SKUs that are stripe payable?  Should there be a flag in the macaroon stating valid payment options?

File: payment/order.go [hunk 7]
```diff
@@ -140,3 +143,61 @@ func CreateOrderItemFromMacaroon(sku string, quantity int) (*OrderItem, error) {
...
+	}
+	return data
+}
+
+func ConvertPriceToUnitAmount(price decimal.Decimal) float64 {
+	n, exact := price.Float64()
+	if !exact {
+		fmt.Println("Error typecasting price")
+	}
+	return n * float64(100.00)
```
Comment (husobee) [inline payment/order.go:202]: This is only valid for USD?  Would it make sense to pass in the currency as a param for the future?

File: payment/controllers.go (hunk 5 above)
Comment (husobee) [reply payment/controllers.go:58]: +1, I like `/v1/payments/webhooks/stripe` better too.

**Pull request closed without merging**